### ISSUE TITLE
Questions regarding new client

### ISSUE THREAD

**Issue by AshotN** [ampache/ampache#2161]
So while I build my prototype for a new client I have some questions for the maintainers, and all users in general, about what they expect in the client.

In the current client, if you go on an album and click on the first song it doesn't automatically play the next song. You have to play the album and not the song if you want the queue to be automatic. I don't know if this is intuitive, and I would prefer for a more Spotify style automatic queue.

I would appreciate a list of the mandatory at launch features the client should have, so I can write the appropriate JSON methods or make issues about missing API methods.

Also if anyone here knows a thing or two about design I'd appreciate help, my artistic side lacks. 

For example I couldn't think of a color that represents the song is currently playing, so I vibrant green should act as a FIXME
![image](https://user-images.githubusercontent.com/1350342/68658508-69140700-04ea-11ea-91e8-7a0bdefc1ec4.png)

**Comment by lachlan-00** [comment]
I'd love to see some feedback on this from people.

For the pages with a list of tracks how about a method that looks for a "next" track?
Loop through an album in order and loop back to the start or stop if at the end and an option is set.

That could apply for basically any list really. Give it an array/query/search with the current value to match then find the next item after the current one. This might be doable through something like the playqueue methods in subsonic set per session.
I think using a playqueue would need to happen to allow smartlists to retain their original search

For me i would focus on the user level function. (Playback/music, search, playlist, random from the menu) video is a low priority due to how bad its implemented and statistics are generally just searches.

Positioning it as a player first would get people using it.

**Comment by lachlan-00** [comment]
For me this is my ampache web-ui workflow.
* go to album mashup to see recent/new/trending (Set a default page on login?)
* go to smart lists
* play either "unplayed artists" or "song > 2stars"
* play random songs using the rating filter. (Ignore 1 star artist, album, song)
* rate while playing but only songs

In the background (on my server) i have the following scripts.
* flag all 5 star objects and 5 star all flagged objects
* when a song is 5 stars or flagged, flag+5star the album and the artist if not set as well
* when a song is set to 3,4,5 stars do a query to find the highest rating for the album and set. Then set the artist to the highest rated album value.

Basically now that i'm set up the admin functions are not important to me.

I kind of don't want a lot of admin api access in the api right away but some things missing from the api might be.
* check for update. (And update)
* get/set preferences (user and admin)
* uploads
* shares
* live streams / radio
* channels (but i kind of dont like them and they're disabled by default on new installs)
* private messages/server chat (chat is in subsonic api but not regular api)

Something that i dont use but lots of people seem to is localplay/mpd and democratic playback. I also think the shoutbox is a bit useless.

**Comment by AshotN** [comment]
It's great that Ampache has a ton of features, but feature creep is a thing. I would love to prioritize thinks based on usage and potentially exclude things that are just not being used at all. For example private messages are really not a priority.

For queue, are you suggesting the server figure out the queue? I was going to make it client side, but there are a few advantages if it's handled by the server. Such as allowing two separate clients to be in sync and control each other. Like if I play a song on desktop but I want to control it all from my phone.

I figure once the client is at a usable state we can provide it to users as an opt-in, and for whatever feature is missing on the new client we can redirect to the old one.

I also try to get really annoying things out of the way first, I'm probably going to do keyboard shortcuts next for example, just because adding them down the line ends of being a huge headache in refactoring. 

@SUTJael had made some mock-up images here https://github.com/ampache/ampache/issues/359#issuecomment-53078744, and I would love if either @SUTJael or anyone else could contribute in the design aspect. I'll write all the CSS, but I'm not to good with conceptualizing something good looking. Which I think is Ampache's critical weak point.

**Comment by lachlan-00** [comment]
There are a few ways to handle a playqueue but they seem to all fit in to "per user" or "per session".

A user handshakes and a session is created.
This can then be used to save a playqueue for the user/device. This isn't saved outside the loaded webplayer but could be considered current default behaviour.

The server has the tables for it because of subsonic so it would be easy to do. It could also be a client option on what to do. 
Api::play_queue(type='user'|'session')

**Comment by AshotN** [comment]
Hmm, we'll we should decide so the client can work with that in mind. We can make the queue generate on the server and send it to the client but that would make the client stop working offline, and if we ever have something like cached music on the client we would still need to inform the server on play and get a queue.

**Comment by lachlan-00** [comment]
I think the queue is managed by the client but it can be saved/retrieved from the server.

* Allow Retrieve from server on login / menu item
* Save on play state / track change (remove finished tracks from the list?)
* Clear on client will clear on server
* option to save or clear on logout

**Comment by AshotN** [comment]
I'm going to implement it client side, and if we decide to create syncing with server later I should be able to just send the data for the API to save.

**Comment by lachlan-00** [comment]
Good idea. I've been making a ton of little fixes this week trying to finalise everything then I'll get onto testing out this

**Comment by kuzi-moto** [comment]
> For example I couldn't think of a color that represents the song is currently playing, so I vibrant green should act as a FIXME

I've done some research on design for websites/apps and it seems typical to select a pallet of several colors and stick to it throughout your design. In reborn, this seems to be the bright yellow color from the logo, though the usage is inconsistent.

Color can always be changed later on, I'm building my interface in greyscale first as I found it's too distracting to deal with when overall design/layout is more important.

**Comment by AshotN** [comment]
Ya I hate colors. I stick with what I call Ampache yellow/orange. But I don't want the color for song playing and hover to be the same for example. 

I also hate dealing with queue logic. I had done this once before, and it wasn't perfect. This time no better, I'm hoping if this project succeeds we can refactor it.

**Comment by AshotN** [comment]
I didn't want to make a new issue for this, but @lachlan-00 you ever considered [GraphQL](https://graphql.org). I've personally never used it, but I hear great things about it. I was thinking maybe it can help revamp the API with less work.

**Comment by lachlan-00** [comment]
that looks awesome, i really wrote a python script to run through the api but it would be good to do something better.

**Comment by AshotN** [comment]
I think GraphQL could really clean the API up as well. I haven't used it personally, but from what I understand, we could grab as much data as we need without creating very specific API points. It be a project, but it might be worth doing this over the JSON API in it's current standing. As this will practically replace it. 

BTW, have you had a chance to build the React client yourself. I added Queuing and the ability to browser playlists. So it's semi-usable in it's bare-bones state.

**Comment by lachlan-00** [comment]
This week I've spent about 30 hours in testing + my regular 38hour week so I haven't tried anything yet.

But the first thing I'm going to do is install this because I'm sick of fixing my own bugs. ;)

**Comment by Rhythm-of-Time** [comment]
Keep up the great work! Btw. in 4.0.2 Song Editor it is not possible to Add New... Artist or Album.

**Comment by AshotN** [comment]
You should probably open a new issue in regards to that.

**Comment by Rhythm-of-Time** [comment]
@AshotN You are right. I will open a new issue. Just thought that lachlan-00 is busy fixing anyway, and he seems to be very active...
Regarding this thread's topic, I'd also like to have a massively simplified user interface. Maybe the best approach would be theme based, so via CSS certain items could be shown or hidden. That would require the possibility to pre-select a theme based on Access level - everybody else than Adminstrators or Catalog Manager would see a reduced version of the interface (this is partly already the case regarding Admin features). I would be grateful for suggestions how to implement this. Maybe somebody will also tell me this is not possible because...? Still I believe that even if some changes to the templates should be necessary to achieve more configurability (let's call it a Backend and a Frontend version) - it would be faster than building a separate client.

**Comment by kuzi-moto** [comment]
@Rhythm-of-Time you could probably just implement the same logic as the sidebar uses to save the state of the expanded/collapsed menu items.

Alternatively, there could be more options in the interface configuration settings.

**Comment by AshotN** [comment]
Frankly the reason I'm advocating for a complete re-write is because the current client(interface, whatever you want to call it) is horribly outdated, cluttered, and a real hassle to work on. 

The Ampache API is powerful enough, and getting better thanks to @lachlan-00, to be able to be the primary way to fetch data. Without needing PHP code in templates files. 

React is also way faster than the current way we are doing things. We can utilize new technologies with a new client. Sometimes it's easier and smarter to start over than to keep patching holes in a sinking ship. 

I implore you to try out my prototype [client/interface](https://github.com/AshotN/ampache/tree/ReactClient). While it's missing most the functionality of the current client, and it's queuing system doesn't work terribly well. It demonstrates that we can build a new client that uses the API, and make it an enjoyable experience for the user. 

One of the major concerns people online seem to have regarding Ampache is it's archaic web interface. While mostly functional, it is not very intuitive and littered with legacy code and unmaintained dependencies. 

Frankly I want to finish the new client and see it shipped with Ampache, but unless people actually want it and will actually use it, it's not worth building it.

**Comment by wagnered** [comment]
@AshotN,

I don't currently have a system to install it on.

Would it be possible to set up a public demo as defined in the ampache Google group. Invite members to comment. Make it easier for your fellow developers and other users to give it a try.  

> If you build it they will come. (Field of Dreams, 1989)

If you believe in your product, don't wait for approval. If you get the messaging system working, the user could leave an immediate comment while trying the demo.

I did a web search for the word "ampache" and got a "ton" of leads for which you could invite users to try your interface: blogs, etc.  BTW: A web host service bragged that they could provide > 95.5% uptime for ampache servers.

@lachlan-00 is there a way to make issues private or only visible by a "Chosen few" (read 'active developers')?  I think we should be able to comment and discuss the more technical design issues without involving the general public who may misinterpret what they read.  Just a suggestion.




-------- Original message --------
Subject: Re: [ampache/ampache] Questions regarding new client (#2161)

**Comment by AshotN** [comment]
I’ll create a guest user on my server and post it here. I don’t know how well it will work on mobile for now. So if you can use it on Desktop preferably. 

Will update soon!

Also perhaps we should have a developers IRC

**Comment by Rhythm-of-Time** [comment]
**@AshotN,** 
thank you for (re)posting a link to your prototype's repository. Yesterday I looked for it and didn't find it, only a quite old version of ampache_react. I agree with you that a complete redesign would be better, so I definitely will have a look at it.

I need a quick solution for a project though, and I think a node.js server is required for React, which we can't run on our shared host. Apart from that, would there still be a performance gain by running a React client from one server and fetching data via API from another server running Ampache? 

I am looking for ways to customize the user interface for the needs of my project - and not sure it can be done by CSS alone as for instance altering the menu structure for a more elegant use of filtering (e.g. genre, mood, instruments) would require changes not only in the templates, but also the database. Lot of work to be done... Maybe your solution will be quicker, I am curious to see it online!

@wagnered 
Being an Ampache user for two weeks only, I did not mean to offend anybody and excuse myself to experienced developers and long-time contributors who sure know better what the future concept of Ampache should be.

**Comment by wagnered** [comment]
@Rhythm-of-Time,

Please, no need to apologize. It is I who should apologize if I appear to be pushing the design in a certain direction.  That is not my intent.  Whatever the future of Ampache, it should be based on what I believe is the best design principals.  I don't care too much about the details.  I will push for the most efficient and effective architecture (MVC, modularity and use of plug-ins). For example, I would like w3.CSS to be used instead of standard CSS.  It is more responsive than standard css and no jquery or JavaScript is need.  Also w3.css is well designed for inline use. But no one else is willing to even try it out. ( ; ' >

null

**Comment by AshotN** [comment]
@Rhythm-of-Time Node is not required. The client compiles our to JavaScript and HTML. Nothing is required to run the client, but you do need node and yarn to build the client. 

@wagnered I’m sorry I’m not sure as that CSS you’re referring to. I’m well acquainted with regular CSS and it’s capable of doing anything really. Can you post what you’re referring to

**Comment by Rhythm-of-Time** [comment]
**@AshotN** So I can build the client on my Windows machine and copy it to a folder on the webserver?

**Comment by Rhythm-of-Time** [comment]
@AshotN I mean a folder connected to a subdomain, for instance. Something like client.server.tld while Ampache is running on ampache.server.tld...

**Comment by wagnered** [comment]
At https://www.w3schools.com/w3css/  you'll find info on w3.css. 

I apologize for generalizing.

I took a look at your repo and I have a few questions.  I also got the URL from the .env file and went there. Of course I got your login page.

I assume you have some functionality implemented.  I got lost at main.tsx.  I assume that when the login button is clicked, the username and password will be posted to the top level index.php? 

Is CSS applied to the login page: I couldn't see the raw data?



-------- Original message --------
Subject: Re: [ampache/ampache] Questions regarding new client (#2161)

**Comment by AshotN** [comment]
https://ampache.hego.co/newclient is my server using the newclient. You can use `test:test` to login.

So I try to organize things a bit in the client. So let's say you are accessing a page like like `https://.../newclient/artist/38`. This request gets routed through `router.tsx`. 

```                               
<Route
    exact
    path='/artist/:artistID'
    render={(props) => (
        <ArtistView
            {...props}
            user={this.state.user}
        />
    )}
/>
```

`router.tsx` returns the `ArtistView` which is imported from `'./Views/Artist`.

The `ArtistView` file is a function that returns JSX, which is what React then converts to HTML for the browser down the line. 

The CSS of the site is written in Stylus, and is in the `stylus` folder. The files need to be cleaned up a bit. But Stylus is really helpful in writing CSS as we can easily compartmentalize files, declare variables, use functions, etc. 

Regarding W3.CSS, it looks like it's made by W3Schools, which isn't held by a high regard these days. Also the project seems unmaintained, https://github.com/vitorlans/w3-css.

If you're saying we should use something like Bootstrap or W3.CSS or one of the many competitors. I really never liked front-end templates. I always felt it's harder to have to keep up with their docs and components and what not then to just build what I need when I need it. Also I think a lot of people suck at CSS because the refuse to write any. I get that, CSS is sometimes an annoying language, but frankly I'm confident in my CSS if someone can design what they want in Photoshop.

**Comment by wagnered** [comment]
> Https://github.com/vitorlans/w3-css

Why would you want to use a third-party project? Never mind, I understand your point.

I tried accessing the URL. The first time I got "connection timeout".  Additional tries, I got "unable to connect".  The problem could be that my android system is using an old version of SSL/TLS.  Usually when accessing a secure site, I get a warning and must ignore the certificate to access the site.  I didn't get that from your URL.



-------- Original message --------
Subject: Re: [ampache/ampache] Questions regarding new client (#2161)

**Comment by wagnered** [comment]
Or maybe your system is down because yesterday I could access https://ampache.hego.co/ and now I can't.null

**Comment by wagnered** [comment]
The path to index.html in /newclient/index.php seems to be incorrect.null

**Comment by AshotN** [comment]
https://ampache.hego.co/newclient/ seems to work for me, and I'm at out of my home network currently.

Perhaps it's TLS being an issue? Would you like me to enable port 80 access?

**Comment by a7medo778** [comment]
i created my own front end client here
shiafm.com

it doesn't have admin stuff, but for frontend consuption it has been working great

**Comment by kuzi-moto** [comment]
@a7medo778 I can't understand any of it, but it does look nice! Do you have the source available?

**Comment by a7medo778** [comment]
> @a7medo778 I can't understand any of it, but it does look nice! Do you have the source available?

haha yes its all in arabic, i created out of this ionic4 template https://store.enappd.com/product/ionic-4-spotify-app-starter/

if you think its viable, i can always create a server nutural version (enter your server info and your live)

**Comment by wagnered** [comment]
I was finally able to make a connection and got the following error when logging in:

> TypeError: e.toString(...).padStart is not a function.

null

**Comment by AshotN** [comment]
> I was finally able to make a connection and got the following error when logging in:
> TypeError: e.toString(...).padStart is not a function.
> null

What version of Android are you running? Or what browser, it's possible the JavaScript is not compatible, and I'll try to see if I can build a compatible version.

**Comment by AshotN** [comment]
@a7medo778 If it's a product you paid for, I don't think you can share it with us for this project. Unless the author of said product permitted it.

**Comment by wagnered** [comment]
This android system is old 2.3.  The only browser I can install that comes close to working is flashfox version 45.5.1.  I'm thinking this is a modified clone of Firefox:  Mainly because of the version number and for add-ons, it goes to the Mozilla add-on site.

Please don't waste too much time on it.  
 
What do you think the learning curve is? How fast could a developer become proficient in the tools and language(s)  necessary to create/modify the client logic?



-------- Original message --------
Subject: Re: [ampache/ampache] Questions regarding new client (#2161)

**Comment by AshotN** [comment]
Well JavaScript is not a hard language at all. It does have some funny quirks but I'd expect a lot of web developers to have dealt with JavaScript already. 

Stylus is simple, as it's 90% CSS.

TypeScript add's types to JavaScript, it's not really to hard to use, and I decent IDE makes it pretty useful.

React is probably the hardest part, I try to keep my files organized so future developers and I myself can understand what's happening and edit things easier. But it does have a learning curve. Anyone who would be interested in writing the Views and Components would need to read the React docs and get some understanding on it. 

But a lot of the action still remains in the back-end, the API especially. So I think if there are devs who want to help but don't care to use JavaScript. Then they could help with the PHP in the back-end.

**Comment by wagnered** [comment]
If you are using the window.fetch API, the earliest Firefox to impleent it is 39.  It won't work on android 2.3. I managed to find Firefox version 36 which works better than any of the other browsers I could try.

I wonder what it would take to upgrade my phone to a more recent android version. ( ; ' >



-------- Original message --------
Subject: Re: [ampache/ampache] Questions regarding new client (#2161)

**Comment by kuzi-moto** [comment]
> I wonder what it would take to upgrade my phone to a more recent android version. ( ; ' >

If you're using a semi popular phone, chances are pretty good that there's a custom ROM available that's a higher version of Android than the OEM provided.

**Comment by Rhythm-of-Time** [comment]
Your client prototype on https://ampache.hego.co/newclient/ works - I can navigate, but it doesn't play audio. I was scratching my head a while looking for a play icon before I tried the right mouse button. Still the webplayer would not load any songs after clicking on play in the context menu. 

A local install from your repository https://github.com/AshotN/ampache/tree/ReactClient works fine - I am not sure which parts are still to come or intentionally left out, but it plays audio. Somehow I am used to the Play, Add to playlist etc. buttons, though these have very small icons in Reborn. Could a right click be handled on mobile devices? But it's a bit early for these questions.

I built the client with the source code changed to use our own Ampache and put it on the server, addressed CORS issues by adding the following code at the top of /server/json.server.php:
// try to get rid of * by replacing with own server
header('Access-Control-Allow-Origin: *');
header('Access-Control-Allow-Headers: X-Requested-With');

Now I am stuck. When I try to login to the "newclient", I always get Error: "Error: Missing Auth Key"
handleLogin router.tsx:58
index.tsx:20:20

Access control is enabled, there are the default ACLs for API/RPC, and I generated an API Key for every user in this testing environment. What am I possibly missing? 

Ampache is 4.03 release.

**Comment by AshotN** [comment]
Odd I can play music from different computers from my server. What browser and audio format are you trying?

A lot of the functionality is in right-click context menus. So if you right-click on a song you should be able to add to playlist. It's missing on the search page for now. But for songs a simple click should start playing. I would look in the developer console for errors if it's not. 

Your error is coming from the server, I would use developer console to see what's happening. I would try using the `handshake` method in the API manually. You can look at the `logic/Auth.ts` file for reference. 

"Play Next" is for adding to the queue(which is a work in progress). The home page shows your current queue, in a very ugly temporary manner. 

The space key will also play/pause the song. If you are using Chromium/Chrome media keys should also work, next/previous song isn't working yet, but play/pause does.

**Comment by kuzi-moto** [comment]
> A lot of the functionality is in right-click context menus.

Personally, I am not a huge fan of sites hijacking right-click, though I don't know how everyone else feels about it. Also as mentioned the interface will still have to supply ways to work with mobile devices.

**Comment by lachlan-00** [comment]
Firefox doesn't let you do much from right click either. it's generally a browser function rather than a site function.

**Comment by wagnered** [comment]
@lachlan-00

Browsers have a default context menu which can be replaced by HTML/JavaScript from the the server.  Depending on the "context" below the mouse, the browser's default menu can be permitted to show.

@kuzi-moto I like the use of context menus.  For the above reason, the display doesn't need to be cluttered up.
 
@AshotN,  I would like to see the use of more modal overlay windows, especially for editing/viewing single items. They can be dismissed and the window repainted instantly without having to reload the page.

Also, could you address kuzi-moto's (and my) concern about browsers on mobile and tablet devices?  How responsive or adaptive will the css be to the various devices of many varied display sizes.

In the next few days I might have temporary access to a computer to try your test site.  But I'm already very convinced you are headed in the right direction.

I have more questions and suggestions but they can wait 'til later.



-------- Original message --------
Subject: Re: [ampache/ampache] Questions regarding new client (#2161)

**Comment by lachlan-00** [comment]
I have an easy example of why right click can be a problem.
Firefox blocks using right click for copy/paste. In cockpit the menu shows but doesn't actually paste. (I have to use shift+insert instead.)

https://github.com/cockpit-project/cockpit/issues/2009
![image](https://user-images.githubusercontent.com/1305249/70013153-7771b200-15c2-11ea-96f4-9d7949b13c76.png)

**Comment by Rhythm-of-Time** [comment]
**@AshotN** 
Stupid me... it was the right-click that confused me. Your live site works like my local build - but at first I was looking for a songs page, and then for play icons. Then I found the right-click context menu and tried "Play next" and "Add to Playlist" (where I selected the "Test" list). Nothing happened. In "Playlists" I found only the Test playlist by admin, checking the top-right menu item (user?) Playlist gave a 404. That's where I gave up checking the live site. The webplayer only loads a song once you left-click a song or find the other context menu I discovered today on the Random Albums - it is not implemented in the Artists View though, where I had tried it first. Btw., I'm on Firefox.

I won't use right-clicking in my project, so I'll try to modify your approach, once I get it working on our server...  I tried the handshake manually like this (values for timestamp and auth taken from the output of digestMessage):

/server/json.server.php?action=handshake&user=${test}&timestamp=${1575311137}&auth=${84eefff5e7c86548b3d5db1262d0672de09b92de0c88adc26eabe10f68741e29}&version=350001`

SyntaxError: JSON.parse: unexpected character at line 1 column 1 of the JSON data
json.server.php.json:
<?xml version="1.0" encoding="UTF-8" ?>
<root>
	<error code="401"><![CDATA[Received Invalid Handshake - Incorrect username or password]]></error>

That is odd, because username and password are correct. Ampache log reads:
2019-12-03 02:49:11 [ampache] (access.class) -> check_network rpc: ip matched ACL 79.192.248.181 
2019-12-03 02:49:11 [ampache] (api.class) -> Handshake Attempt, IP:79.192.248.181 User:${test} Version:350001` 
2019-12-03 02:49:11 [ampache] (api.class) -> Login Attempt, IP:79.192.248.181 Time: 1575337501 User:${test} () Auth:${ab442cd3be5d2afcf1d0e975f91bdf596c913bff81d1a02f6281dc0469da7cc9} 
2019-12-03 02:49:11 [ampache] (api.class) -> Login Failed, unable to match passphrase 

If I attempt a normal login via newclient, the logs say:
2019-12-03 03:00:44 [ampache] (access.class) -> check_network rpc: ip matched ACL 79.192.248.181 
2019-12-03 03:00:44 [ampache] (api.class) -> Handshake Attempt, IP:79.192.248.181 User:test Version:350001 
2019-12-03 03:00:44 [ampache] (api.class) -> Login Attempt, IP:79.192.248.181 Time: 1575338403 User:test (4) Auth:213489ab16ff832d5f41b939ce0e63ce3d9e32514ba20d824635cc3cc74e0047 
2019-12-03 03:00:44 [ampache] (access.class) -> check_network rpc: ip matched ACL 79.192.248.181 
2019-12-03 03:00:44 [ampache] (session.class) -> Unable to read session from key f26eca0a9ae9762c1bdbc2f240a778ea no data found 
2019-12-03 03:00:44 [ampache] (session.class) -> Session created: 69eba4981db89ab955fb70503f83b40c 
2019-12-03 03:00:44 [ampache] (api.class) -> Login Success, passphrase matched 
2019-12-03 03:00:44 [ampache] (access.class) -> check_network rpc: ip matched ACL 79.192.248.181 
2019-12-03 03:00:44 [ampache] (api.class) -> Handshake Attempt, IP:79.192.248.181 User:test Version:350001 
2019-12-03 03:00:44 [ampache] (api.class) -> Login Attempt, IP:79.192.248.181 Time: 1575338403 User:test (4) Auth:213489ab16ff832d5f41b939ce0e63ce3d9e32514ba20d824635cc3cc74e0047 
2019-12-03 03:00:44 [ampache] (access.class) -> check_network rpc: ip matched ACL 79.192.248.181 
2019-12-03 03:00:44 [ampache] (session.class) -> Unable to read session from key f26eca0a9ae9762c1bdbc2f240a778ea no data found 
2019-12-03 03:00:44 [ampache] (dba.class) -> Error_query SQL: INSERT INTO `session` (`id`, `username`, `ip`, `type`, `agent`, `value`, `expire`, `geo_latitude`, `geo_longitude`, `geo_name`) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?) 
2019-12-03 03:00:44 [ampache] (dba.class) -> Error_query MSG: ["23000",1062,"Duplicate entry '69eba4981db89ab955fb70503f83b40c' for key 'PRIMARY'"] 
2019-12-03 03:00:44 [ampache] (dba.class) -> Error_query SQL: INSERT INTO `session` (`id`, `username`, `ip`, `type`, `agent`, `value`, `expire`, `geo_latitude`, `geo_longitude`, `geo_name`) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?) 
2019-12-03 03:00:44 [ampache] (dba.class) -> Error_query MSG: ["23000",1062,"Duplicate entry '69eba4981db89ab955fb70503f83b40c' for key 'PRIMARY'"] 
2019-12-03 03:00:44 [ampache] (dba.class) -> Error_query SQL: INSERT INTO `session` (`id`, `username`, `ip`, `type`, `agent`, `value`, `expire`, `geo_latitude`, `geo_longitude`, `geo_name`) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?) 
2019-12-03 03:00:44 [ampache] (dba.class) -> Error_query MSG: ["23000",1062,"Duplicate entry '69eba4981db89ab955fb70503f83b40c' for key 'PRIMARY'"] 
2019-12-03 03:00:44 [ampache] (dba.class) -> Error_query SQL: INSERT INTO `session` (`id`, `username`, `ip`, `type`, `agent`, `value`, `expire`, `geo_latitude`, `geo_longitude`, `geo_name`) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?) 
2019-12-03 03:00:44 [ampache] (dba.class) -> Error_query MSG: ["23000",1062,"Duplicate entry '69eba4981db89ab955fb70503f83b40c' for key 'PRIMARY'"] 
2019-12-03 03:00:44 [ampache] (session.class) -> Session creation failed 
2019-12-03 03:00:44 [ampache] (api.class) -> Login Success, passphrase matched 

I looks like login is happening, but the Auth key is still missing.

**Comment by AshotN** [comment]
As far as context menu hijacking goes, I feel modern websites and apps kind of have made users used to it. I'm completely fine with adding a option to disable the context-menu hijacking. Context menus are for making things more intuitive and faster to use, I don't think they should be the only way to do things. 

For mobile, iOS is not playing any audio, and I sadly don't have an Android anymore to test with. I do ideally want to make the site work on mobile. But Safari on iOS and generally all mobile browsers suck, so desktop will probably have more features than mobile. It will definitely be a step-up from the Reborn interface as far as mobile goes.

Overlays are cool, but the only one I am really using right now is for adding to playlist. I don't know if I am really happy with the code I used to make the overlay. It works, but I'm not sure if it's the cleanest code. I'm not terribly worried about switching between views because React-Router doesn't do a traditional 'refresh', so it really doesn't matter. The API call takes the same amount of time one way or another. 

@Rhythm-of-Time I'm not sure what's causing your error. It's definitely server side, perhaps one of the files is out of date. I would update all your files with mainline Ampache, and then follow the guide on my Fork for the files that need to be replaced. Once Ampache has a JSON API mainline, that will make life easier. 

Thank you all for your input!

**Comment by AshotN** [comment]
I pushed a [commit](https://github.com/AshotN/ampache/commit/3a709ada1ceca5d9395d2bf907861803c22c6ef2) that allows you to access the custom context-menu without right click.
![image](https://user-images.githubusercontent.com/1350342/70019950-0232bc80-1583-11ea-9b72-0c7ac1ed6026.png)

**Comment by wagnered** [comment]
After giving thought the problem, I think I understand the situation and believe there is a misunderstanding if you believe that the browser should paste directly into the terminal.   When you use the keyboard(shift insert), the focus is on the terminal.  When you use the mouse, the focus is on the topmost app. In this case the browser. The browser is limited to it's window area for cut/copy/paste operations. The browser probably isn't even "aware" that the terminal process is running.  There would need to be some very nontrivial (re)programming for interprocess communications between the browser and terminal.

If I understand the situation correctly, the Firefox browser is operating as designed.  If I don't then ignore all the above. ( ; ' >null

**Comment by Rhythm-of-Time** [comment]
@AshotN -
Finally got it working. My error: I hkept omitting https:// when calling the app- and there was no ReWrite rule set... Thank you for your help! 

Holding the left mouse button is a mobile-friendly approach. I would prefer it to right-click. Maybe some visual clue is needed - perhaps a little burger?

**Comment by AshotN** [comment]
Glad to hear it, the client is pretty unfinished and definitely unpolished. If you want to make any aesthetic changes, I'm all for it.

**Comment by lachlan-00** [comment]
i wonder if you could make this into a dropin jukebox style player? Maybe use something like playlist_generate to create a simple streaming interface to just play things. with a quick play selection of:
'recent', 'forgotten', 'unplayed', 'random'

Putting that in %site%/jukebox would be a pretty killer feature and pretty simple compared to replacing the entire interface

**Comment by lachlan-00** [comment]
what put this in my head was the way your sidebar is done. it's essentially already there with just the sidebar and some different menu options.

**Comment by AshotN** [comment]
We can always add such a feature. I’m fine with leaving Reborn as default. Can have 

`/newclient/jukebox`

**Comment by teroneko** [comment]
Hey I would like to join the party and would really like to enjoy to help to build a new fresh ampache web interface. Any interests in contribution? @AshotN I wrote already to you (email), because you are the one who uses the right tools, great. :)

**Comment by lachlan-00** [comment]
@teroneko feel free to do or work on anything you want to. there are a whole heap of people working this year who never contributed before and it's really exciting to see

**Comment by AshotN** [comment]
Sorry for the inactivity. Work has picked up, and I just haven't been able to find the time to work on the client. Hopefully soon I can make more progress.

**Comment by mitchray** [comment]
Closing due to React client going on hiatus: https://github.com/ampache/ampache/issues/2360#issuecomment-1342288895
